src/celutil/logger.h

```diff
@@ -0,0 +1,123 @@
+#pragma once
+
+#include <fmt/core.h>
+#include <fmt/printf.h>
+
+#include <celcompat/filesystem.h>
+template <>
+struct fmt::formatter<fs::path> : formatter<std::string>
+{
+    template <typename FormatContext>
+    auto format(const fs::path &path, FormatContext &ctx)
+    {
+        return formatter<std::string>::format(path.string(), ctx);
+    }
+};
+
+namespace celestia
+{
+namespace util
+{
+
+enum class Level
+{
+    Error,
+    Warning,
+    Info,
+    Verbose,
+    Debug,
+};
+
+class Logger
+{
+ public:
+    Logger() = default;
+    Logger(Level level) :
+        m_level(level)
+    {}
+
+    void setLevel(Level level)
+    {
+        m_level = level;
+    }
+
+    template <typename... Args> inline void
+    debug(const char *format, const Args&... args) const;
+
+    template <typename... Args> inline void
+    info(const char *format, const Args&... args) const;
+
+    template <typename... Args> inline void
+    verbose(const char *format, const Args&... args) const;
+
+    template <typename... Args> inline void
+    warn(const char *format, const Args&... args) const;
+
+    template <typename... Args> inline void
+    error(const char *format, const Args&... args) const;
+
+    template <typename... Args> void
+    log(Level, const char *format, const Args&... args) const;
+
+    template <typename... Args> void
+    log2(Level, const char *format, const Args&... args) const;
+
+    static Logger* g_logger;
+
+ private:
+    void vlog(Level level, fmt::string_view format, fmt::format_args args) const;
+
+    void vlog2(Level level, fmt::string_view format, fmt::printf_args args) const;
+
+    Level m_level { Level::Info };
+};
+
+template <typename... Args> void
+Logger::log(Level level, const char *format, const Args&... args) const
+{
+    if (level <= m_level)
+        vlog(level, fmt::string_view(format), fmt::make_format_args(args...));
+}
+
+template <typename... Args> void
+Logger::debug(const char *format, const Args&... args) const
+{
+    log(Level::Debug, format, args...);
+}
+
+template <typename... Args> inline void
+Logger::info(const char *format, const Args&... args) const
+{
+    log(Level::Info, format, args...);
+}
+
+template <typename... Args> inline void
+Logger::verbose(const char *format, const Args&... args) const
+{
+    log(Level::Verbose, format, args...);
+}
+
+template <typename... Args> inline void
+Logger::warn(const char *format, const Args&... args) const
+{
+    log(Level::Warning, format, args...);
+}
+
+template <typename... Args> inline void
+Logger::error(const char *format, const Args&... args) const
+{
+    log(Level::Error, format, args...);
+}
+
+template <typename... Args> void
+Logger::log2(Level level, const char *format, const Args&... args) const
```

Comment:
I'm wondering if we should take the opportunity to update the printf-style strings to the new fmt syntax (with apologies to the translators), to reduce the API surface.

Reply:
i will do this later, i want to writer a script to update translations first

Reply:
or i can start working on the script now and defer this pr

Reply:
Ok no worries, would rather get this one merged first I think.

Reply:
ok, i decided to got with the proposal and updated some languages